src/lxd.py

```diff
@@ -0,0 +1,388 @@
+            logger.exception("Failed to stop LXD instance")
+            raise LxdError(f"Unable to stop LXD instance {self.name}") from err
+
+    def delete(self, wait: bool = False) -> None:
+        """Delete the LXD instance.
+
+        Args:
+            wait: Whether to wait until the LXD instance stopped before returning.
+
+        Raises:
+            LxdException: Unable to delete the LXD instance.
+        """
+        try:
+            self._pylxd_instance.delete(wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            logger.exception("Failed to delete LXD instance")
+            raise LxdError(f"Unable to delete LXD instance {self.name}") from err
+
+    def execute(self, cmd: list[str], cwd: Optional[str] = None) -> Tuple[int, IO, IO]:
+        """Execute a command within the LXD instance.
+
+        Exceptions are not raise if command execution failed. Caller should check the exit code and
+        stderr for failures.
+
+        Args:
+            cmd: Commands to be executed.
+            cwd: Working directory to execute the commands.
+
+        Returns:
+            Tuple containing the exit code, stdout, stderr.
+        """
+        lxc_cmd = ["/snap/bin/lxc", "exec", self.name]
+        if cwd:
+            lxc_cmd += ["--cwd", cwd]
+
+        lxc_cmd += ["--"] + cmd
+
+        result = secure_run_subprocess(lxc_cmd)
+        return (result.returncode, io.BytesIO(result.stdout), io.BytesIO(result.stderr))
+
+
+class LxdInstanceManager:
+    """LXD instance manager."""
+
+    def __init__(self, pylxd_client: pylxd.Client):
+        """Construct the LXD instance manager.
+
+        Args:
+            pylxd_client: Instance of pylxd.Client.
+        """
+        self._pylxd_client = pylxd_client
+
+    def all(self) -> list[LxdInstance]:
+        """Get list of LXD instances.
+
+        Raises:
+            LxdException: Unable to get all LXD instance.
+
+        Returns:
+            List of LXD instances.
+        """
+        try:
+            return [
+                LxdInstance(instance.name, instance)
+                for instance in self._pylxd_client.instances.all()
+            ]
+        except pylxd.exceptions.LXDAPIException as err:
+            logger.exception("Failed to get all LXD instance")
+            raise LxdError("Unable to get all LXD instances") from err
+
+    def create(self, config: LxdInstanceConfig, wait: bool) -> LxdInstance:
+        """Create a LXD instance.
+
+        Args:
+            config: Configuration for the LXD instance.
+            wait: Whether to wait until the LXD instance created before returning.
+
+        Raises:
+            LxdException: Unable to get all LXD instance.
+
+        Returns:
+            The created LXD instance.
+        """
+        try:
+            pylxd_instance = self._pylxd_client.instances.create(config=config, wait=wait)
+            return LxdInstance(config["name"], pylxd_instance)
```

Comment:
Is it possible that "name"  key does not exist for some reason?

Reply:
Currently, it is not possible as the name is passed in on Runner class creation.
But checking can be added to the `__init__` method to ensure a valid name is set.

Reply:
Minor comments